// devexpress-gantt/dist/lib/src/Commands/Task/RemoveTaskCommand.ts
import { ConfirmationType } from "../../Dialogs/DialogEnums";
import { ConfirmationDialogParameters } from "../../Dialogs/DialogParameters/ConfirmationDialogParameters";
import { TaskRemovingArguments } from "../../Model/Events/Task/TaskRemovingArguments";
import { RemoveDependencyHistoryItem } from "../../Model/History/HistoryItems/Dependency/RemoveDependencyHistoryItem";
import { DeassignResourceHistoryItem } from "../../Model/History/HistoryItems/ResourceAssignment/DeassignResourceHistoryItem";
import { RemoveTaskHistoryItem } from "../../Model/History/HistoryItems/Task/RemoveTaskHistoryItem";
import { SimpleCommandState } from "../SimpleCommandState";
import { TaskCommandBase } from "./TaskCommandBase";

export class RemoveTaskCommand extends TaskCommandBase {
    public execute(id: string, confirmRequired: boolean = true, isApiCall: boolean = false, isUpdateParentTaskRequired: boolean = true, historyItem?: RemoveTaskHistoryItem): boolean {
        this.isApiCall = isApiCall;
        this.isUpdateParentTaskRequired = isUpdateParentTaskRequired;
        if(confirmRequired) {
            this.control.commandManager.showConfirmationDialog.execute(new ConfirmationDialogParameters(ConfirmationType.TaskDelete,
                () => { this.executeInternal(id, historyItem); }));
            return false;
        }
        return super.execute(id, historyItem);
    }
    protected executeInternal(id: string, historyItem?: RemoveTaskHistoryItem): boolean {
        id = id || this.control.currentSelectedTaskID;
        const item = this.control.viewModel.findItem(id);
        const task = item ? item.task : this.control.viewModel.tasks.getItemById(id);
        const args = new TaskRemovingArguments(task);
        this.modelManipulator.dispatcher.notifyTaskRemoving(args);
        if(args.cancel)
            return false;
        this.control.history.beginTransaction();
        this.control.viewModel.beginUpdate();
        const removeTaskHistoryItem = historyItem || new RemoveTaskHistoryItem(this.modelManipulator);
        removeTaskHistoryItem.addTask(id);
        const childTasks = this.control.viewModel.tasks.items.filter(t => t.parentId === id);
        childTasks.forEach(t => new RemoveTaskCommand(this.control).execute(t.internalId, false, true, false, removeTaskHistoryItem));

        let dependencies = this.control.viewModel.dependencies.items.filter(d => d.predecessorId === id || d.successorId === id);
        if(dependencies.length)
            if(this.control.settings.editing.allowDependencyDelete) {
                dependencies = dependencies.filter(d => childTasks.filter(c => c.internalId === d.predecessorId || c.internalId === d.successorId).length === 0);
                dependencies.forEach(d => removeTaskHistoryItem.add(new RemoveDependencyHistoryItem(this.modelManipulator, d.internalId)));
            }
            else
                return false;

        const assignments = this.control.viewModel.assignments.items.filter(a => a.taskId === id);
        assignments.forEach(a => {
            if(this.modelManipulator.dispatcher.fireResourceUnassigning(a))
                removeTaskHistoryItem.add(new DeassignResourceHistoryItem(this.modelManipulator, a.internalId));
        });
        if(!historyItem)
            this.history.addAndRedo(removeTaskHistoryItem);
        const parent = this.control.viewModel.findItem(task.parentId);
        if(this.isUpdateParentTaskRequired)
            super.updateParent(parent);
        this.control.history.endTransaction();
        this.control.viewModel.endUpdate();
        return true;
    }
    isEnabled(): boolean {
        const gantt = this.control;
        const selectedItem = gantt.viewModel.findItem(gantt.currentSelectedTaskID);
        const result = super.isEnabled() && (!!selectedItem && selectedItem.selected || this.isApiCall);
        return result;
    }
    getState(): SimpleCommandState {
        const state = super.getState();
        const gantt = this.control;
        state.visible = state.visible && gantt.settings.editing.allowTaskDelete;
        return state;
    }
    private isUpdateParentTaskRequired: boolean;
}
