// devexpress-gantt/dist/lib/src/Dialogs/TaskEditDialog.ts
import { ResourceCollection } from "../Model/Collections/ResourceCollection";
import { TaskEditDialogShowingArguments } from "../Model/Events/Dialogs/TaskEditDialogShowingArguments";
import { ResourceAssigningArguments } from "../Model/Events/ResourceAssignment/ResourceAssigningArguments";
import { AssignResourceHistoryItem } from "../Model/History/HistoryItems/ResourceAssignment/AssignResourceHistoryItem";
import { DeassignResourceHistoryItem } from "../Model/History/HistoryItems/ResourceAssignment/DeassignResourceHistoryItem";
import { TaskEndHistoryItem } from "../Model/History/HistoryItems/Task/Properties/TaskEndHistoryItem";
import { TaskProgressHistoryItem } from "../Model/History/HistoryItems/Task/Properties/TaskProgressHistoryItem";
import { TaskStartHistoryItem } from "../Model/History/HistoryItems/Task/Properties/TaskStartHistoryItem";
import { TaskTitleHistoryItem } from "../Model/History/HistoryItems/Task/Properties/TaskTitleHistoryItem";
import { Task } from "../Model/Entities/Task";
import { DialogBase } from "./DialogBase";
import { TaskEditParameters } from "./DialogParameters/TaskEditParameters";
import { SimpleCommandState } from "../Commands/SimpleCommandState";

export class TaskEditDialogCommand extends DialogBase<TaskEditParameters> {
    onBeforeDialogShow(params: TaskEditParameters): boolean {
        return this.modelManipulator.dispatcher.raiseTaskTaskEditDialogShowing(
            params,
            (args: TaskEditDialogShowingArguments) => {
                const newValues = args.values;
                params.start = newValues.start;
                params.end = newValues.end;
                params.progress = newValues.progress;
                params.title = newValues.title;
                params.readOnlyFields = args.readOnlyFields;
                params.hiddenFields = args.hiddenFields;
            }
        );
    }
    applyParameters(newParameters: TaskEditParameters, oldParameters: TaskEditParameters): boolean {
        this.history.beginTransaction();

        const success = this.control.modelManipulator.dispatcher.raiseTaskMultipleUpdating(
            this.control.viewModel.tasks.getItemById(oldParameters.id),
            newParameters,
            (newValues) => {
                newParameters.title = newValues.title ? newValues.title : "";
                newParameters.progress = newValues.progress;
                newParameters.start = typeof newValues.start === "string" ? new Date(newValues.start) : newValues.start as Date || new Date(0);
                newParameters.end = typeof newValues.end === "string" ? new Date(newValues.end) : newValues.end as Date || new Date(0);

            });

        if(success) {
            if(newParameters.title !== oldParameters.title)
                this.history.addAndRedo(new TaskTitleHistoryItem(this.modelManipulator, oldParameters.id, newParameters.title));
            if(newParameters.progress !== oldParameters.progress)
                this.history.addAndRedo(new TaskProgressHistoryItem(this.modelManipulator, oldParameters.id, newParameters.progress));
            if(newParameters.end.getTime() < newParameters.start.getTime())
                newParameters.end = newParameters.start;
            if(newParameters.start !== oldParameters.start) {
                this.history.addAndRedo(new TaskStartHistoryItem(this.modelManipulator, oldParameters.id, newParameters.start));
                if(this.control.isValidateDependenciesRequired())
                    this.control.validationController.moveStartDependTasks(oldParameters.id, oldParameters.start);
            }
            if(newParameters.end !== oldParameters.end) {
                this.history.addAndRedo(new TaskEndHistoryItem(this.modelManipulator, oldParameters.id, newParameters.end));
                if(this.control.isValidateDependenciesRequired())
                    this.control.validationController.moveEndDependTasks(oldParameters.id, oldParameters.end);
            }
        }

        for(let i = 0; i < newParameters.assigned.length; i++) {
            const resource = oldParameters.assigned.getItemById(newParameters.assigned.getItem(i).internalId);
            if(!resource) {
                const resourceId = newParameters.assigned.getItem(i).internalId;
                const taskId = oldParameters.id;
                const args = new ResourceAssigningArguments(resourceId, taskId);
                this.modelManipulator.dispatcher.notifyResourceAssigning(args);
                if(!args.cancel)
                    this.history.addAndRedo(new AssignResourceHistoryItem(this.modelManipulator, args.resourceId, args.taskId));
            }
        }
        for(let i = 0; i < oldParameters.assigned.length; i++) {
            const assigned = oldParameters.assigned.getItem(i);
            const resource = newParameters.assigned.getItemById(assigned.internalId);
            if(!resource) {
                const assignment = this.control.viewModel.assignments.items.filter(assignment => assignment.resourceId === assigned.internalId && assignment.taskId === oldParameters.id)[0];
                if(this.modelManipulator.dispatcher.fireResourceUnassigning(assignment))
                    this.history.addAndRedo(new DeassignResourceHistoryItem(this.modelManipulator, assignment.internalId));
            }
        }
        const updateParents = newParameters.start !== oldParameters.start || newParameters.end !== oldParameters.end || newParameters.progress !== oldParameters.progress;
        if(success && updateParents)
            this.validationController.updateParentsIfRequired(oldParameters.id);
        this.history.endTransaction();
        return false;
    }
    createParameters(options: Task): TaskEditParameters {
        options = options || this.control.viewModel.tasks.getItemById(this.control.currentSelectedTaskID);
        const param = new TaskEditParameters();
        param.id = options.internalId;
        param.title = options.title;
        param.progress = options.progress;
        param.start = options.start;
        param.end = options.end;
        param.assigned = this.control.viewModel.getAssignedResources(options);
        param.resources = new ResourceCollection();
        param.resources.addRange(this.control.viewModel.resources.items);
        param.showResourcesDialogCommand = this.control.commandManager.showResourcesDialog;
        param.showTaskEditDialogCommand = this.control.commandManager.showTaskEditDialog;
        param.enableEdit = this.isTaskEditEnabled();
        param.enableRangeEdit = this.isTaskRangeEditEnabled(options);
        param.isValidationRequired = this.control.isValidateDependenciesRequired();
        param.getCorrectDateRange = (taskId: any, startDate: Date, endDate: Date) => { return this.control.validationController.getCorrectDateRange(taskId, startDate, endDate); };

        return param;
    }
    isTaskEditEnabled(): boolean {
        const settings = this.control.settings;
        return settings.editing.enabled && settings.editing.allowTaskUpdate;
    }
    isTaskRangeEditEnabled(task: Task): boolean {
        return !this.control.viewModel.isTaskToCalculateByChildren(task.internalId);
    }
    isEnabled(): boolean {
        const gantt = this.control;
        const selectedItem = gantt.viewModel.findItem(gantt.currentSelectedTaskID);
        return (!!selectedItem && selectedItem.selected) || this.isApiCall;
    }
    getState(): SimpleCommandState {
        const state = super.getState();
        state.visible = state.visible && !this.control.taskEditController.dependencyId;
        return state;
    }
    getDialogName(): string {
        return "TaskEdit";
    }
}
