// devexpress-gantt/dist/lib/src/Model/History/HistoryItems/Dependency/InsertDependencyHistoryItem.ts
import { Dependency } from "../../../Entities/Dependency";
import { DependencyType } from "../../../Entities/Enums";
import { ModelManipulator } from "../../../Manipulators/ModelManipulator";
import { HistoryItem } from "../HistoryItem";

export class InsertDependencyHistoryItem extends HistoryItem {
    dependency: Dependency;
    predecessorId: string;
    successorId: string;
    type: DependencyType;

    constructor(modelManipulator: ModelManipulator, predecessorId: string, successorId: string, type: DependencyType) {
        super(modelManipulator);
        this.predecessorId = predecessorId;
        this.successorId = successorId;
        this.type = type;
    }
    public redo(): void {
        this.dependency = this.modelManipulator.dependency.insertDependency(this.predecessorId, this.successorId, this.type,
            this.dependency ? this.dependency.internalId : null);
    }
    public undo(): void {
        this.modelManipulator.dependency.removeDependency(this.dependency.internalId);
    }
}
