// devexpress-gantt/dist/lib/src/Model/Manipulators/Dependency/DependencyManipulator.ts
import { Dependency } from "../../Entities/Dependency";
import { DependencyType } from "../../Entities/Enums";
import { BaseManipulator } from "../BaseManipulator";

export class TaskDependencyManipulator extends BaseManipulator {
    insertDependency(predecessorId: string, successorId: string, type: DependencyType, id?: string): Dependency {
        const dependency = this.viewModel.dependencies.createItem();
        dependency.predecessorId = predecessorId;
        dependency.successorId = successorId;
        dependency.type = type;
        if(id)
            dependency.internalId = id;
        dependency.id = dependency.internalId;
        this.viewModel.dependencies.add(dependency);
        this.dispatcher.notifyDependencyInserted(this.getObjectForDataSource(dependency), id => dependency.id = id, this.getErrorCallback());
        this.viewModel.updateVisibleItemDependencies();
        this.renderHelper.recreateConnectorLineElement(dependency.internalId, true);
        return dependency;
    }
    removeDependency(dependencyId: string): Dependency {
        const dependency = this.viewModel.dependencies.getItemById(dependencyId);
        this.viewModel.dependencies.remove(dependency);
        this.dispatcher.notifyDependencyRemoved(dependency.id, this.getErrorCallback(), this.viewModel.getDependencyObjectForDataSource(dependency));
        this.viewModel.updateVisibleItemDependencies();
        this.renderHelper.recreateConnectorLineElement(dependency.internalId);
        return dependency;
    }
    private getObjectForDataSource(dependency: Dependency) {
        const predecessor = this.viewModel.tasks.getItemById(dependency.predecessorId);
        const successor = this.viewModel.tasks.getItemById(dependency.successorId);
        const dependencyObject = {
            id: dependency.id,
            predecessorId: predecessor.id,
            successorId: successor.id,
            type: dependency.type
        };

        return dependencyObject;
    }
}
